File: src/utils/dataUtils.ts
import { DataColumn, Dataset, DataType, DataProfile } from '@/types/data';

/**
 * Detects the data type of a column based on its values
 * @param values Array of values to analyze
 * @returns The detected data type
 */
export const detectColumnType = (values: any[]): DataType => {
  // Remove null/undefined values for analysis
  const cleanValues = values.filter(v => v !== null && v !== undefined && v !== '');
  
  if (cleanValues.length === 0) return 'text';
  
  // Check if it's numeric
  const numericCount = cleanValues.filter(v => !isNaN(Number(v)) && isFinite(Number(v))).length;
  const numericRatio = numericCount / cleanValues.length;
  
  if (numericRatio > 0.8) return 'numeric';
  
  // Check if it's datetime
  const dateCount = cleanValues.filter(v => {
    const date = new Date(v);
    return !isNaN(date.getTime()) && v.toString().match(/\d{1,4}[-/]\d{1,2}[-/]\d{1,4}|\d{1,2}[-/]\d{1,2}[-/]\d{2,4}/);
  }).length;
  const dateRatio = dateCount / cleanValues.length;
  
  if (dateRatio > 0.6) return 'datetime';
  
  // Check if it's categorical (limited unique values relative to total)
  const uniqueValues = new Set(cleanValues).size;
  const uniqueRatio = uniqueValues / cleanValues.length;
  
  if (uniqueRatio < 0.1 && uniqueValues < 20) return 'categorical';
  
  return 'text';
};

/**
 * Analyzes a dataset and returns structured data with column information
 * @param rows Raw data rows
 * @param filename Name of the source file
 * @returns Analyzed dataset structure
 */
export const analyzeDataset = (rows: Record<string, any>[], filename: string): Dataset => {
  if (rows.length === 0) {
    return {
      name: filename,
      rows: [],
      columns: [],
      totalRows: 0,
      totalColumns: 0,
      memoryUsage: 0
    };
  }
  
  const columnNames = Object.keys(rows[0]);
  const columns: DataColumn[] = columnNames.map(name => {
    const values = rows.map(row => row[name]);
    const type = detectColumnType(values);
    const missingCount = values.filter(v => v === null || v === undefined || v === '').length;
    const uniqueCount = new Set(values).size;
    
    return {
      name,
      type,
      values,
      missingCount,
      uniqueCount
    };
  });
  
  const memoryUsage = JSON.stringify(rows).length;
  
  return {
    name: filename,
    rows,
    columns,
    totalRows: rows.length,
    totalColumns: columnNames.length,
    memoryUsage
  };
};

/**
 * Formats bytes into human-readable format
 * @param bytes Number of bytes
 * @returns Formatted string
 */
export const formatBytes = (bytes: number): string => {
  if (bytes === 0) return '0 Bytes';
  const k = 1024;
  const sizes = ['Bytes', 'KB', 'MB', 'GB'];
  const i = Math.floor(Math.log(bytes) / Math.log(k));
  return parseFloat((bytes / Math.pow(k, i)).toFixed(2)) + ' ' + sizes[i];
};

/**
 * Returns an emoji icon for a given data type
 * @param type Data type
 * @returns Emoji string
 */
export const getColumnIcon = (type: DataType): string => {
  const iconMap: Record<DataType, string> = {
    numeric: '🔢',
    categorical: '🏷️',
    datetime: '📅',
    text: '📝'
  };
  return iconMap[type] || '❓';
};

/**
 * Detects outliers in numeric data using IQR method
 * @param values Array of numeric values
 * @returns Array of outlier values
 */
export const detectOutliers = (values: number[]): number[] => {
  if (values.length === 0) return [];
  
  const sorted = [...values].sort((a, b) => a - b);
  const q1 = sorted[Math.floor(sorted.length * 0.25)];
  const q3 = sorted[Math.floor(sorted.length * 0.75)];
  const iqr = q3 - q1;
  const lowerBound = q1 - 1.5 * iqr;
  const upperBound = q3 + 1.5 * iqr;
  
  return values.filter(v => v < lowerBound || v > upperBound);
};

/**
 * Calculates Pearson correlation coefficient between two numeric arrays
 * @param x First array
 * @param y Second array
 * @returns Correlation coefficient (-1 to 1)
 */
export const calculateCorrelation = (x: number[], y: number[]): number => {
  if (x.length !== y.length || x.length === 0) return 0;
  
  const n = x.length;
  const sumX = x.reduce((sum, val) => sum + val, 0);
  const sumY = y.reduce((sum, val) => sum + val, 0);
  const sumXY = x.reduce((sum, val, i) => sum + val * y[i], 0);
  const sumXX = x.reduce((sum, val) => sum + val * val, 0);
  const sumYY = y.reduce((sum, val) => sum + val * val, 0);
  
  const numerator = n * sumXY - sumX * sumY;
  const denominator = Math.sqrt((n * sumXX - sumX * sumX) * (n * sumYY - sumY * sumY));
  
  return denominator === 0 ? 0 : numerator / denominator;
};

/**
 * Generates a comprehensive data profile for quality assessment
 * @param dataset Dataset to analyze
 * @returns Data profile with quality metrics
 */
export const generateDataProfile = (dataset: Dataset): DataProfile => {
  const profile: DataProfile = {
    overview: {
      totalRows: dataset.totalRows,
      totalColumns: dataset.totalColumns,
      memoryUsage: dataset.memoryUsage,
      completeness: 0
    },
    quality: {
      duplicates: 0,
      missingValues: 0,
      outliers: 0,
      inconsistentTypes: 0,
      textIssues: 0
    },
    columns: dataset.columns.map(col => ({
      name: col.name,
      type: col.type,
      uniqueCount: col.uniqueCount,
      missingCount: col.missingCount,
      completeness: ((dataset.totalRows - col.missingCount) / dataset.totalRows) * 100
    }))
  };
  
  // Calculate overall completeness
  const totalCells = dataset.totalRows * dataset.totalColumns;
  const totalMissing = dataset.columns.reduce((sum, col) => sum + col.missingCount, 0);
  profile.overview.completeness = ((totalCells - totalMissing) / totalCells) * 100;
  
  // Calculate quality metrics
  profile.quality.missingValues = totalMissing;
  
  // Detect duplicates
  const uniqueRows = new Set(dataset.rows.map(row => JSON.stringify(row)));
  profile.quality.duplicates = dataset.totalRows - uniqueRows.size;
  
  // Count outliers in numeric columns
  dataset.columns.forEach(col => {
    if (col.type === 'numeric') {
      const values = col.values.filter(v => v !== null && v !== undefined && !isNaN(Number(v))).map(Number);
      const outliers = detectOutliers(values);
      profile.quality.outliers += outliers.length;
    }
  });
  
  return profile;
};

/**
 * Debounces a function call
 * @param func Function to debounce
 * @param wait Wait time in milliseconds
 * @returns Debounced function
 */
export const debounce = <T extends (...args: any[]) => any>(
  func: T,
  wait: number
): ((...args: Parameters<T>) => void) => {
  let timeout: NodeJS.Timeout;
  return (...args: Parameters<T>) => {
    clearTimeout(timeout);
    timeout = setTimeout(() => func(...args), wait);
  };
};

/**
 * Safely parses a numeric value
 * @param value Value to parse
 * @returns Parsed number or null if invalid
 */
export const safeParseNumber = (value: any): number | null => {
  if (value === null || value === undefined || value === '') return null;
  const parsed = Number(value);
  return isNaN(parsed) || !isFinite(parsed) ? null : parsed;
};

/**
 * Calculates basic statistics for numeric data
 * @param values Array of numeric values
 * @returns Statistics object
 */
export const calculateBasicStats = (values: number[]) => {
  if (values.length === 0) return null;
  
  const sorted = [...values].sort((a, b) => a - b);
  const sum = values.reduce((acc, val) => acc + val, 0);
  const mean = sum / values.length;
  
  const variance = values.reduce((acc, val) => acc + Math.pow(val - mean, 2), 0) / values.length;
  const std = Math.sqrt(variance);
  
  const median = sorted.length % 2 === 0
    ? (sorted[sorted.length / 2 - 1] + sorted[sorted.length / 2]) / 2
    : sorted[Math.floor(sorted.length / 2)];
  
  return {
    mean,
    median,
    std,
    variance,
    min: sorted[0],
    max: sorted[sorted.length - 1],
    count: values.length
  };
};

File: python_core/data_utils.py
"""
Core data processing and analysis utilities.
"""
import json
import math
import statistics
from typing import List, Dict, Any, Optional, Tuple, Union
from collections import Counter
import re
from datetime import datetime

from .types import DataType, Dataset, DataColumn, ColumnStats, DataProfile, DataQualityIssues


def detect_column_type(values: List[Any]) -> DataType:
    """
    Detects the data type of a column based on its values.
    
    Args:
        values: List of values to analyze
        
    Returns:
        The detected data type
    """
    # Remove null/undefined values for analysis
    clean_values = [v for v in values if v is not None and v != '']
    
    if not clean_values:
        return DataType.TEXT
    
    # Check if it's numeric
    numeric_count = 0
    for v in clean_values:
        try:
            float(v)
            if math.isfinite(float(v)):
                numeric_count += 1
        except (ValueError, TypeError):
            pass
    
    numeric_ratio = numeric_count / len(clean_values)
    
    if numeric_ratio > 0.8:
        return DataType.NUMERIC
    
    # Check if it's datetime
    date_count = 0
    date_patterns = [
        r'\d{1,4}[-/]\d{1,2}[-/]\d{1,4}',
        r'\d{1,2}[-/]\d{1,2}[-/]\d{2,4}'
    ]
    
    for v in clean_values:
        str_v = str(v)
        # Try parsing as date
        try:
            datetime.fromisoformat(str_v.replace('/', '-'))
            date_count += 1
            continue
        except ValueError:
            pass
        
        # Check date patterns
        for pattern in date_patterns:
            if re.search(pattern, str_v):
                date_count += 1
                break
    
    date_ratio = date_count / len(clean_values)
    
    if date_ratio > 0.6:
        return DataType.DATETIME
    
    # Check if it's categorical (limited unique values relative to total)
    unique_values = len(set(clean_values))
    unique_ratio = unique_values / len(clean_values)
    
    if unique_ratio < 0.1 and unique_values < 20:
        return DataType.CATEGORICAL
    
    return DataType.TEXT


def analyze_dataset(rows: List[Dict[str, Any]], filename: str) -> Dataset:
    """
    Analyzes a dataset and returns structured data with column information.
    
    Args:
        rows: Raw data rows
        filename: Name of the source file
        
    Returns:
        Analyzed dataset structure
    """
    if not rows:
        return Dataset(
            name=filename,
            rows=[],
            columns=[],
            total_rows=0,
            total_columns=0,
            memory_usage=0
        )
    
    column_names = list(rows[0].keys())
    columns = []
    
    for name in column_names:
        values = [row.get(name) for row in rows]
        data_type = detect_column_type(values)
        missing_count = sum(1 for v in values if v is None or v == '')
        unique_count = len(set(v for v in values if v is not None and v != ''))
        
        columns.append(DataColumn(
            name=name,
            type=data_type,
            values=values,
            missing_count=missing_count,
            unique_count=unique_count
        ))
    
    memory_usage = len(json.dumps(rows).encode('utf-8'))
    
    return Dataset(
        name=filename,
        rows=rows,
        columns=columns,
        total_rows=len(rows),
        total_columns=len(column_names),
        memory_usage=memory_usage
    )


def format_bytes(bytes_count: int) -> str:
    """
    Formats bytes into human-readable format.
    
    Args:
        bytes_count: Number of bytes
        
    Returns:
        Formatted string
    """
    if bytes_count == 0:
        return '0 Bytes'
    
    k = 1024
    sizes = ['Bytes', 'KB', 'MB', 'GB']
    i = int(math.floor(math.log(bytes_count) / math.log(k)))
    
    return f"{round(bytes_count / math.pow(k, i), 2)} {sizes[i]}"


def get_column_icon(data_type: DataType) -> str:
    """
    Returns an emoji icon for a given data type.
    
    Args:
        data_type: Data type
        
    Returns:
        Emoji string
    """
    icon_map = {
        DataType.NUMERIC: '🔢',
        DataType.CATEGORICAL: '🏷️',
        DataType.DATETIME: '📅',
        DataType.TEXT: '📝'
    }
    return icon_map.get(data_type, '❓')


def detect_outliers(values: List[float]) -> List[float]:
    """
    Detects outliers in numeric data using IQR method.
    
    Args:
        values: List of numeric values
        
    Returns:
        List of outlier values
    """
    if not values:
        return []
    
    sorted_values = sorted(values)
    n = len(sorted_values)
    
    q1 = sorted_values[int(n * 0.25)]
    q3 = sorted_values[int(n * 0.75)]
    iqr = q3 - q1
    
    lower_bound = q1 - 1.5 * iqr
    upper_bound = q3 + 1.5 * iqr
    
    return [v for v in values if v < lower_bound or v > upper_bound]


def calculate_correlation(x: List[float], y: List[float]) -> float:
    """
    Calculates Pearson correlation coefficient between two numeric arrays.
    
    Args:
        x: First array
        y: Second array
        
    Returns:
        Correlation coefficient (-1 to 1)
    """
    if len(x) != len(y) or len(x) == 0:
        return 0.0
    
    n = len(x)
    sum_x = sum(x)
    sum_y = sum(y)
    sum_xy = sum(xi * yi for xi, yi in zip(x, y))
    sum_xx = sum(xi * xi for xi in x)
    sum_yy = sum(yi * yi for yi in y)
    
    numerator = n * sum_xy - sum_x * sum_y
    denominator = math.sqrt((n * sum_xx - sum_x * sum_x) * (n * sum_yy - sum_y * sum_y))
    
    return numerator / denominator if denominator != 0 else 0.0


def safe_parse_number(value: Any) -> Optional[float]:
    """
    Safely parses a numeric value.
    
    Args:
        value: Value to parse
        
    Returns:
        Parsed number or None if invalid
    """
    if value is None or value == '':
        return None
    
    try:
        parsed = float(value)
        return parsed if math.isfinite(parsed) else None
    except (ValueError, TypeError):
        return None


def calculate_basic_stats(values: List[float]) -> Optional[ColumnStats]:
    """
    Calculates basic statistics for numeric data.
    
    Args:
        values: List of numeric values
        
    Returns:
        Statistics object or None if no valid data
    """
    if not values:
        return None
    
    try:
        mean_val = statistics.mean(values)
        median_val = statistics.median(values)
        std_val = statistics.stdev(values) if len(values) > 1 else 0.0
        variance_val = statistics.variance(values) if len(values) > 1 else 0.0
        min_val = min(values)
        max_val = max(values)
        
        # Calculate quartiles
        sorted_values = sorted(values)
        n = len(sorted_values)
        q1 = sorted_values[int(n * 0.25)]
        q3 = sorted_values[int(n * 0.75)]
        quartiles = [min_val, q1, median_val, q3, max_val]
        
        return ColumnStats(
            mean=mean_val,
            median=median_val,
            std=std_val,
            variance=variance_val,
            min_val=min_val,
            max_val=max_val,
            quartiles=quartiles
        )
    except statistics.StatisticsError:
        return None


def generate_data_profile(dataset: Dataset) -> DataProfile:
    """
    Generates a comprehensive data profile for quality assessment.
    
    Args:
        dataset: Dataset to analyze
        
    Returns:
        Data profile with quality metrics
    """
    # Calculate overall completeness
    total_cells = dataset.total_rows * dataset.total_columns
    total_missing = sum(col.missing_count for col in dataset.columns)
    completeness = ((total_cells - total_missing) / total_cells) * 100 if total_cells > 0 else 0
    
    # Detect duplicates
    unique_rows = set(json.dumps(row, sort_keys=True) for row in dataset.rows)
    duplicates = dataset.total_rows - len(unique_rows)
    
    # Count outliers in numeric columns
    outliers_count = 0
    for col in dataset.columns:
        if col.type == DataType.NUMERIC:
            numeric_values = [safe_parse_number(v) for v in col.values]
            valid_values = [v for v in numeric_values if v is not None]
            outliers = detect_outliers(valid_values)
            outliers_count += len(outliers)
    
    # Count text issues
    text_issues = 0
    for col in dataset.columns:
        if col.type in [DataType.TEXT, DataType.CATEGORICAL]:
            text_values = [str(v) for v in col.values if isinstance(v, str)]
            has_issues = any(
                v != v.strip() or  # leading/trailing spaces
                re.search(r'\s{2,}', v)  # multiple spaces
                for v in text_values
            )
            if has_issues:
                text_issues += 1
    
    quality = DataQualityIssues(
        duplicates=duplicates,
        missing_values=total_missing,
        outliers=outliers_count,
        inconsistent_types=0,  # Would need more complex analysis
        text_issues=text_issues
    )
    
    overview = {
        'total_rows': dataset.total_rows,
        'total_columns': dataset.total_columns,
        'memory_usage': dataset.memory_usage,
        'completeness': completeness
    }
    
    columns_info = []
    for col in dataset.columns:
        col_completeness = ((dataset.total_rows - col.missing_count) / dataset.total_rows) * 100 if dataset.total_rows > 0 else 0
        columns_info.append({
            'name': col.name,
            'type': col.type.value,
            'unique_count': col.unique_count,
            'missing_count': col.missing_count,
            'completeness': col_completeness
        })
    
    return DataProfile(
        overview=overview,
        quality=quality,
        columns=columns_info
    )


def debounce(func, wait_time: float):
    """
    Decorator to debounce function calls.
    
    Args:
        func: Function to debounce
        wait_time: Wait time in seconds
        
    Returns:
        Debounced function
    """
    import threading
    import time
    
    def debounced(*args, **kwargs):
        def call_func():
            time.sleep(wait_time)
            func(*args, **kwargs)
        
        # Cancel previous timer if exists
        if hasattr(debounced, '_timer'):
            debounced._timer.cancel()
        
        # Start new timer
        debounced._timer = threading.Timer(wait_time, call_func)
        debounced._timer.start()
    
    return debounced